File: src/bmsx/render/vdp/skybox.ts
import type { VdpHostOutput } from '../../machine/devices/vdp/vdp';
import { SKYBOX_FACE_COUNT } from '../../machine/devices/vdp/contracts';
import type { GameView } from '../gameview';

export function commitVdpSkyboxViewState(view: GameView, output: VdpHostOutput): void {
	if (!output.skyboxEnabled) {
		view.skyboxFaceUvRects = null;
		view.skyboxFaceTextpageBindings = null;
		view.skyboxFaceSizes = null;
		return;
	}
	const faceUvRects = view.skyboxFaceUvRects ?? (view.skyboxFaceUvRects = new Float32Array(SKYBOX_FACE_COUNT * 4));
	const faceTextpageBindings = view.skyboxFaceTextpageBindings ?? (view.skyboxFaceTextpageBindings = new Int32Array(SKYBOX_FACE_COUNT));
	const faceSizes = view.skyboxFaceSizes ?? (view.skyboxFaceSizes = new Int32Array(SKYBOX_FACE_COUNT * 2));
	for (let index = 0; index < SKYBOX_FACE_COUNT; index += 1) {
		const sample = output.skyboxSamples[index]!;
		const uvBase = index * 4;
		faceUvRects[uvBase + 0] = sample.source.srcX / sample.surfaceWidth;
		faceUvRects[uvBase + 1] = sample.source.srcY / sample.surfaceHeight;
		faceUvRects[uvBase + 2] = sample.source.width / sample.surfaceWidth;
		faceUvRects[uvBase + 3] = sample.source.height / sample.surfaceHeight;
		faceTextpageBindings[index] = sample.slot;
		const sizeBase = index * 2;
		faceSizes[sizeBase + 0] = sample.source.width;
		faceSizes[sizeBase + 1] = sample.source.height;
	}
}

File: src/bmsx_cpp/render/vdp/skybox.cpp
#include "render/vdp/skybox.h"

#include "machine/devices/vdp/vdp.h"
#include "render/gameview.h"

namespace bmsx {

void commitVdpSkyboxViewState(GameView& view, const VDP::VdpHostOutput& output) {
	if (!output.skyboxEnabled) {
		view.skyboxRenderReady = false;
		view.skyboxFaceUvRects = {};
		view.skyboxFaceTextpageBindings = {};
		view.skyboxFaceSizes = {};
		return;
	}
	for (size_t index = 0; index < SKYBOX_FACE_COUNT; ++index) {
		const VDP::ResolvedBlitterSample& sample = (*output.skyboxSamples)[index];
		const VDP::BlitterSource& source = sample.source;
		const size_t uvBase = index * 4u;
		view.skyboxFaceUvRects[uvBase + 0u] = static_cast<f32>(source.srcX) / static_cast<f32>(sample.surfaceWidth);
		view.skyboxFaceUvRects[uvBase + 1u] = static_cast<f32>(source.srcY) / static_cast<f32>(sample.surfaceHeight);
		view.skyboxFaceUvRects[uvBase + 2u] = static_cast<f32>(source.width) / static_cast<f32>(sample.surfaceWidth);
		view.skyboxFaceUvRects[uvBase + 3u] = static_cast<f32>(source.height) / static_cast<f32>(sample.surfaceHeight);
		view.skyboxFaceTextpageBindings[index] = static_cast<i32>(sample.slot);
		const size_t sizeBase = index * 2u;
		view.skyboxFaceSizes[sizeBase + 0u] = static_cast<i32>(source.width);
		view.skyboxFaceSizes[sizeBase + 1u] = static_cast<i32>(source.height);
	}
	view.skyboxRenderReady = true;
}

} // namespace bmsx
